packages/authorization-server/src/services/tokenService.ts

```
@@ -0,0 +1,434 @@
+        throw platformStateValidationFailed();
+      }
+
+      const { limitReached, ...rateLimiterStatus } =
+        await redisRateLimiter.rateLimitByOrganization(key.consumerId, logger);
+      if (limitReached) {
+        return {
+          limitReached: true,
+          token: undefined,
+          rateLimitedTenantId: key.consumerId,
+          rateLimiterStatus,
+        };
+      }
+
+      // TODO: match otherwise doesn't work somehow
+      if (key.clientKind === clientKindTokenStates.consumer) {
+        const token = await tokenGenerator.generateInteropConsumerToken({
+          sub: jwt.payload.sub,
+          audience: key.eServiceState.audience,
+          purposeId: key.purposeId,
+          tokenDurationInSeconds: 10,
+        });
+
+        await publishAudit({
+          producer,
+          generatedToken: token,
+          key,
+          clientAssertion: jwt,
+          correlationId,
+          fileManager,
+          logger,
+        });
+
+        return {
+          limitReached: false,
+          token,
+          rateLimiterStatus,
+        };
+      } else if (key.clientKind === clientKindTokenStates.api) {
+        const token = await tokenGenerator.generateInteropApiToken({
+          sub: jwt.payload.sub,
+          consumerId: key.consumerId,
+        });
+
+        return {
+          limitReached: false,
+          token,
+          rateLimiterStatus,
+        };
+      } else {
+        throw unexpectedTokenGenerationError();
+      }
+    },
+  };
+}
+
+export type TokenService = ReturnType<typeof tokenServiceBuilder>;
+
+export const retrieveKey = async (
+  dynamoDBClient: DynamoDBClient,
+  pk: TokenGenerationStatesClientKidPurposePK | TokenGenerationStatesClientKidPK
+): Promise<ConsumerKey | ApiKey> => {
+  const input: GetItemInput = {
+    Key: {
+      PK: { S: pk },
+    },
+    TableName: config.tokenGenerationStatesTable,
+  };
+
+  // try {
+  // TODO should we use try/catch in every dynamoDB query?
```

If we avoid this, we could throw the specific errors as shown below. The downside is that any other connection error is thrown as it is.

---

Inside the router, at the higher level `try/catch` all the unhandled errors should be caught as well and turned into `ApiProblem`. We will double-check this during live tests.